### ISSUE TITLE
Request handling time metrics

### ISSUE THREAD

**Issue by rianmcguire** [puma/puma#1541]
tl;dr: Collect per-request threadpool wait time and other interesting timings, and expose them via the Rack environment.

We've just started using Puma in production and we're still in the process of tuning it. We're collecting Puma's backlog metric, as well as measuring the time spent inside our Rack app.

We occasionally have large bursts of requests, and we can see from the backlog that requests are waiting for Puma's threadpool, but we don't have a measure for how long they are waiting. We have haproxy in front of Puma, so we can get the total request time from there, but that's distorted by the time taken for slow clients to send request bodies.

I'm considering modifying Puma to track various times in the request lifecycle and calculate the threadpool wait time for each request, as well as the time spent waiting for the request body. These times could be exposed via Rack environment as, eg. `puma.threadpool_wait` and `puma.request_wait` and then collected by middleware.

Is this a feature the maintainers would be interested in having in Puma?

**Comment by nateberkopec** [comment]
> Is this a feature the maintainers would be interested in having in Puma?

👍 Absolutely.

**Comment by stereobooster** [comment]
> Collect per-request threadpool wait time

If you use `nginx` you can set

```
proxy_set_header X-Request-Start "${msec}";
```

and then you can create Rack middleware which will read this value and calculate difference between this value and current time - this is time between request got to webserver and actually started to execute

[Example of Rack middleware](https://github.com/blacklane/kiev/blob/master/lib/kiev/rack/request_logger.rb#L97)

**Comment by rianmcguire** [comment]
That's a great idea, thanks @stereobooster. Looks like something similar is possible with haproxy:
```
http-request set-header X-Request-Start t=%Ts%
```
I'll still need to subtract the request body read time, but that's an easy wait to measure the socket backlog + threadpool wait times. I was a bit stuck on measuring the time in the socket backlog, since puma stops accepting connections when the pool is full, and there's no socket API for determining how long a connection has been waiting.
